### ISSUE TITLE
Single-page application packaging all frontends together?

### ISSUE THREAD

**Issue by jakhog** [IFRCGo/cbs#580]
Would it be of interest to package all of the frontends form the different bounded contexts together into a sort of "super" single-page application (SPA)?

I don't recall all the details from the discussion about the authentication and authorisation from last week, but I think I remember that the three separate SPAs was causing some extra trouble for us. So if there is no good reason to keep them separate, I think this is a good idea.

Can you provide some input @einari, @Velcrow81, @Virrum, or add anyone else that might have a say in this discussion?

I've been experimenting a bit with the idea just now, and think I have found a way to do it that would work nicely. It keeps the current code and separation between the different frontends as it is today, and still allows running a single frontend alone for development purposes.

The approach I have in mind, can also be used to package other common components/modules that can be imported by the different frontends as well (like the navigation bar), to avoid code-duplication.

---

To give you some idea of what I'm thinking, each of the frontends (and common parts, like the navigation bar) would expose a top-level module (like they already do), that can be packaged using ng-packagr. These packages can be "installed" using npm dependencies, and then imported as normal to be used.

The "super" SPA, would then simply import all the frontend top level modules, and render them into a `<router-outlet>` depending on the URL (as we already do).

---

If you want to see my example, I can put it in a new repo somewhere so you can try it out.

**Comment by Velcrow81** [comment]
Sounds like a good idea to me. What do you think @einari? This will be instead of an iframe then

**Comment by alexanderhoset** [comment]
mmm so the idea is to package as one, but still have separate contexts   ?
Like an import all to navigation and use that to deploy all?

**Comment by jakhog** [comment]
Well, it depends a bit on what you mean by _context_. But yes, they would be separate Angular-modules and don't know about each other, and we keep the frontend code in their own folders as now.

They would however run in the same JavaScript "context" (as there is only one in the browser), so they would have the potential to screw up for each other if we do something wrong. But since we keep the ability to test them separately, I don't think that's a huge problem.

I was actually planning to keep the navigation (and other common things, e.g. a footer, custom controls) as is, and have a new "web-root" that does the import all and deploy. But that's mostly details...

**Comment by beatfactor** [comment]
This is the same thing I had in mind and I have already proposed this solution here: https://github.com/IFRCGo/cbs/issues/403. Using npm packages seems like the best way to go about it, and since the npm packages are versioned, we can choose what/when to package and deploy. Having an SPA makes things a lot easier in respect to things like performance, deployment, and user experience. 

Regarding the navigation, I agree that it should be in the main root template, because that rarely changes, but if we feel like it should, we can make that an npm package as well.

@jakhog Maybe you could send a PR with your proposed solution and then we can try it out in your fork?

**Comment by jakhog** [comment]
Hi @beatfactor, I was looking for you, but didn't know your GitHub name. Thanks for joining in!

With regards to the versioning of packages, this would require us to publish the packages somewhere. That was not really my plan, but maybe it's not a bad idea.

The reason for keeping the navigation as a separate component, would be to keep the ability to run a single frontend during development/testing. The idea was that each frontend is fully self-contained, and able to work on it's own. Then the new super-SPA, simply imports all of them, and has a very simple index.html file (which would be the same as for the other frontends).

I can definitely implement it in my fork! I just wanted to check if there was a reason for not doing it, other than technical difficulties. I'm not really an expert on all these design concepts that has been chosen for CBS, so I don't want to break anything.

**Comment by beatfactor** [comment]
@jakhog yes, we can simply publish them in the public npmjs.org repository, once we have the individual packages. I'm not familiar with ng-packegr, but it seems that it is doing the job.

**Comment by einari** [comment]
Great input. Composition is interesting. The thinking behind the split and eventually also on the front end has been to prevent dependencies and remove friction and let each individual Bounded Context be somewhat autonomous. 

The different Bounded contexts are also not necessarily used by the same user, and in fact not available for all users. So having a circuit breaker in the sense of postback between the different contexts is not a problem. 

I don't think there will be any issues with the iFrame solution with regards to security the way I've done everything now. 

Instead of creating a superweb with all the stuff bundled in one and if we want to avoid iFrame - we could make the navigation a single HTML file and include it as a HTML template in the other Bounded contexts and render it there as an element. 

I'm working on a design for frontend packaging that is inspired by Dockers layering. Hoping it could be the future. 
Until then I suggest we go either for an iFrame solution or a HTML template approach. Those will then still allow the original we set out without creating potential friction and dependencies in build and release as well as runtime.

**Comment by Velcrow81** [comment]
Since @einari is doing a sprint to clue a lot of things together, I think we should wait until after easter to do any changes so he avoids having to hit a moving target. If the resulting iframe or HTML template prove difficult to use with i.e. security er can use this as plan B

**Comment by einari** [comment]
Thanks @Velcrow81 :) 
I think we can pretty much go either way once it's all up and running from a security perspective. 

Would love to have an awesome answer for composition that doesn't freak everyone out like it seems iFrames does :)

**Comment by jakhog** [comment]
Agreed, will keep this on hold then.

Haha, I'm not so concerned about the iframes (albeit I can't deny it feels strange) 😛 

It was more an attempt to make the login/identity/token things that was discussed. So that it could be handled in JS across the frontends, in stead of putting things in cookies or session storage. But if that's working now, then there's no need to fix it 😄

**Comment by beatfactor** [comment]
Well, people are "freaked out" about iframes because there's a general consensus among front-end developers that they should be avoided. It's one of those things that it seems like it's a good idea, but you know there's something wrong with it, and the more you work with iframes, the more problems you find, and then you start thinking that maybe this wasn't such a good idea after all. :)

**Comment by einari** [comment]
I've been brought up in the dev community to think that iframes are the root of all evil. I'm not so convinced anymore for a typical line-of-business application, as this is the case for CBS. From my research I've seen things like O365 using iframes intensively for the purpose of composition of multiple microservices. Also looking at Azure and other services, you'll see it being done. 

I really can't see any evidence of it being absolutely the wrong path right now for this particular solution and similar solutions we're working on. But very willing to stand corrected and be completely wrong about this and for all I know you might be completely right and I'll snap back and dislike iframes - but right now I just want to give it a go. For the security part I'm working on it won't have much of an impact and we've already done the iframe thing - so we're just going to hook it all up over the next couple of days and see if it works or not.
